connectors/sources/outlook.py

```diff
@@ -0,0 +1,944 @@
+        search_query = ",".join(map(str, ldap_domain_name_list))
+
+        has_value_for_normal_users, _, response, _ = self._create_connection.search(
+            search_query,
+            SEARCH_FILTER_FOR_NORMAL_USERS,
+            attributes=["mail"],
+        )
+
+        (
+            has_value_for_admin_users,
+            _,
+            response_for_admin,
+            _,
+        ) = self._create_connection.search(
+            search_query,
+            SEARCH_FILTER_FOR_ADMIN,
+            attributes=["mail"],
+        )
+
+        if not (has_value_for_normal_users or has_value_for_admin_users):
+            raise UsersFetchFailed(
+                "Error while fetching users from Exchange Active Directory."
+            )
+
+        response.extend(response_for_admin)
+        for user in response:
+            yield user
+
+    async def get_user_accounts(self):
+        BaseProtocol.HTTP_ADAPTER_CLS = (
+            RootCAAdapter if self.ssl_enabled else NoVerifyHTTPAdapter
+        )
+
+        credentials = Credentials(
+            username=self.user,
+            password=self.password,
+        )
+        configuration = Configuration(
+            credentials=credentials,
+            server=self.exchange_server,
+            retry_policy=FaultTolerance(max_wait=120),
+        )
+
+        async for user in self.get_users():
+            if "searchResRef" in user["type"]:
+                continue
+
+            user_account = Account(
+                primary_smtp_address=user.get("attributes", {}).get("mail"),
+                config=configuration,
+                access_type=IMPERSONATION,
+            )
+            yield user_account
+
+
+class Office365Users:
+    """Fetch users from Office365 Active Directory"""
+
+    def __init__(self, client_id, client_secret, tenant_id):
+        self.tenant_id = tenant_id
+        self.client_id = client_id
+        self.client_secret = client_secret
+
+    @cached_property
+    def _get_session(self):
+        return aiohttp.ClientSession(raise_for_status=True)
+
+    def _check_errors(self, response):
+        match response.status:
+            case 400:
+                raise UnauthorizedException("Found invalid tenant id value")
+            case 401:
+                raise UnauthorizedException(
+                    "Found invalid client id or client secret value"
+                )
+            case 403:
+                raise Forbidden(
+                    f"Missing permission or something went wrong. Error: {response}"
+                )
+            case 404:
+                raise NotFound(f"Resource Not Found. Error: {response}")
+            case _:
+                raise
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+        skipped_exceptions=UnauthorizedException,
+    )
+    async def _fetch_token(self):
```

Comment:
Can the token be cached?

Reply:
it's used once so i did not make it cached